crates/cli/src/commands/patterns_test.rs

```diff
@@ -276,7 +297,232 @@ pub(crate) async fn run_patterns_test(
         bail!("No testable patterns found. To test a pattern, make sure it is defined in .grit/grit.yaml or a .md file in your .grit/patterns directory.");
     }
     info!("Found {} testable patterns.", testable_patterns.len(),);
-    get_marzano_pattern_test_results(testable_patterns, &libs, arg, flags.into()).await
+    get_marzano_pattern_test_results(testable_patterns, &libs, &arg, flags.clone().into()).await?;
+
+    if arg.watch {
+        enable_watch_mode(arg, flags);
+    }
+    Ok(())
+}
+
+fn enable_watch_mode(cmd_arg: PatternsTestArgs, cmd_flags: GlobalFormatFlags) -> () {
+    let path = Path::new(".grit");
+    // setup debouncer
+    let (tx, rx) = std::sync::mpsc::channel();
+    // notify backend configuration
+    let backend_config = notify::Config::default().with_poll_interval(Duration::from_millis(10));
+    // debouncer configuration
+    let debouncer_config = Config::default()
+        .with_timeout(Duration::from_millis(10))
+        .with_notify_config(backend_config);
+    // select backend via fish operator, here PollWatcher backend
+    let mut debouncer = new_debouncer_opt::<_, notify::PollWatcher>(debouncer_config, tx).unwrap();
+
+    debouncer
+        .watcher()
+        .watch(path, RecursiveMode::Recursive)
+        .unwrap();
+    log::info!("\n[Watch Mode] enabled on path: {}", path.display());
+
+    // event pocessing
+    for result in rx {
+        match result {
+            Ok(event) => {
+                let modified_file_path = event.get(0).unwrap().path.clone();
+                log::info!("\n[Watch Mode] File modified: {:?}", modified_file_path);
+
+                let mut arg = cmd_arg.clone();
+                let flags = cmd_flags.clone();
+                arg.watch = false; //avoid creating infinite watchers
+                tokio::task::spawn(async move {
+                    let affected_patterns_names =
+                        get_affected_patterns(modified_file_path).await.unwrap();
+
+                    info!(
+                        "[Watch Mode] Patterns changed/affected : {:?}\n[Watch Mode] Re-running test for changed/affected patterns...\n",
+                        affected_patterns_names
+                    );
+
+                    let _ = run_patterns_test(arg, flags, Some(affected_patterns_names)).await;
+                });
+            }
+            Err(error) => {
+                log::error!("Error {error:?}")
+            }
+        }
+    }
+}
+
+async fn get_affected_patterns(modified_file_path: PathBuf) -> Result<Vec<String>> {
+    let modified_file_path_string = modified_file_path.into_os_string().into_string().unwrap();
+
+    let (patterns, _) = resolve_from_cwd(&Source::Local).await?;
+    let affected_patterns = collect_testable_patterns(patterns.clone())
```

Comment:
Rebuilding this from scratch every time is inefficient. We should instead just be looking at the *existing* retrieve patterns that were first found in the original top level command, then filtering those down to the ones related to the modified file.

Reply:
addressed in: https://github.com/getgrit/gritql/pull/298#discussion_r1614373584